packages/jsii-pacmak/lib/targets/go/types/go-type-reference.ts

```diff
@@ -74,66 +75,109 @@ export class GoTypeRef {
+   * so that return types of interfaces can correctly be constructucted.
+   */
+  public scopedImplMap(scope: Package) {
+    const nextRef =
+      this.typeMap.array ?? this.typeMap.map ?? this.typeMap.interface;
 
-      return `[]${innerName}`;
+    if (nextRef) {
+      return [nextRef.scopedInterfaceName(scope), nextRef.scopedName(scope)];
     }
 
-    // type is a map
-    if (this.reference.mapOfType) {
+    return [];
+  }
+
+  private scopedTypeName(
+    typeMap: TypeMap,
+    scope: Package,
+    asInterface: boolean,
+    asRef = false,
+  ): string {
+    if (typeMap.primitive) {
+      return typeMap.primitive;
+    } else if (typeMap.array || typeMap.map) {
+      const prefix = typeMap.array ? '[]' : 'map[string]';
       const innerName =
-        new GoTypeRef(this.root, this.reference.mapOfType).scopedTypeName(
+        this.scopedTypeName(
+          (typeMap.array ?? typeMap.map!).typeMap,
           scope,
           asInterface,
           asRef,
         ) ?? 'interface{}';
-      return `map[string]${innerName}`;
-    }
-
-    const prefix = asRef ? '*' : '';
-    const baseName = asInterface ? this.interfaceName : this.name;
-    // type is defined in the same scope as the current one, no namespace required
-    if (scope.packageName === this.namespace && baseName) {
-      // if the current scope is the same as the types scope, return without a namespace
-      return `${prefix}${baseName}`;
-    }
-
-    // type is defined in another module and requires a namespace and import
-    if (baseName) {
-      return `${prefix}${this.namespace}.${baseName}`;
+      return `${prefix}${innerName}`;
+    } else if (typeMap.interface) {
+      const prefix = asRef ? '*' : '';
+      const baseName = asInterface
+        ? typeMap.interface.interfaceName
+        : typeMap.interface.name;
+      // type is defined in the same scope as the current one, no namespace required
+      if (scope.packageName === typeMap.interface.namespace && baseName) {
+        // if the current scope is the same as the types scope, return without a namespace
+        return `${prefix}${baseName}`;
+      }
+
+      // type is defined in another module and requires a namespace and import
+      if (baseName) {
+        return `${prefix}${typeMap.interface.namespace}.${baseName}`;
+      }
+    } else if (typeMap.union) {
+      return 'interface{}';
+    } else if (typeMap.void) {
+      return '';
     }
 
     // type isn't handled
-    log.debug(
-      `Type ${baseName} does not resolve to a known Go type. It is being mapped to "interface{}".`,
+    log.warn(
+      `Type ${typeMap.interface?.name} does not resolve to a known Go type. It is being mapped to "interface{}".`,
     );
+
     return 'interface{}';
   }
 }
+
+interface TypeMap {
+  primitive?: string;
+  array?: GoTypeRef;
+  map?: GoTypeRef;
+  union?: GoTypeRef[];
+  void?: boolean;
+  interface?: GoTypeRef;
+}
```

Comment:
Since `scopedTypeName` will actually only care about ONE of them at a time, it appears to me this type is excessively permissive (it'll allow all the fields to be set, which appears to be illegal).

I'd also make this as immutable as possible to make sure nothing does anything stupid with that.

```suggestion
type TypeMap =
     { readonly primitive?: string; }
  || { readonly array?: GoTypeRef; }
  || { readonly map?: GoTypeRef; }
  || { readonly union?: readonly GoTypeRef[]; }
  || { readonly void?: boolean; }
  || { readonly interface?: GoTypeRef; }
```